File: src/yolo.cpp
#include <fstream>
#include <iostream>
#include <vector>
#include <cstddef>
#include <opencv2/opencv.hpp>

std::vector<std::string> load_class_list() {
  std::vector<std::string> class_list;
  std::ifstream ifs("network/classes_mario.txt");
  std::string line;
  while (getline(ifs, line)) {
    class_list.push_back(line);
  }
  return class_list;
}

void load_net(cv::dnn::Net &net, bool is_cuda) {
  auto result = cv::dnn::readNet("network/yolov5s_mario.onnx");
  if (is_cuda) {
    std::cout << "Attempty to use CUDA\n";
    result.setPreferableBackend(cv::dnn::DNN_BACKEND_CUDA);
    result.setPreferableTarget(cv::dnn::DNN_TARGET_CUDA_FP16);
  } else {
    std::cout << "Running on CPU\n";
    result.setPreferableBackend(cv::dnn::DNN_BACKEND_OPENCV);
    result.setPreferableTarget(cv::dnn::DNN_TARGET_CPU);
  }
  net = result;
}

const std::vector<cv::Scalar> colors = {
    cv::Scalar(255, 255, 0), cv::Scalar(0, 255, 0), cv::Scalar(0, 255, 255),
    cv::Scalar(255, 0, 0)};

const int INPUT_WIDTH = 640;
const int INPUT_HEIGHT = 640;
const float SCORE_THRESHOLD = 0.; // to match with the python implementation
const float NMS_THRESHOLD = 0.4;
const float CONFIDENCE_THRESHOLD = 0.4;

struct Detection {
  int class_id;
  float confidence;
  cv::Rect box;
};

/**
 * @brief fit the input image to the smallest square, starting from the upper left
 */
cv::Mat make_square(const cv::Mat &source, int& out_size) {
  int col = source.cols;
  int row = source.rows;
  out_size = std::max(col, row);
  cv::Mat result = cv::Mat::zeros(out_size, out_size, CV_8UC3);
  source.copyTo(result(cv::Rect(0, 0, col, row)));
  return result;
}

void detect(const cv::Mat &image, cv::dnn::Net &net, std::vector<Detection> &output,
            std::vector<int> &data_record,
            const std::vector<std::string> &className) {
  cv::Mat blob;

  /*--------------------Pre-process--------------------*/
  int pre_blob_image_size;
  // Preprocessing
  cv::dnn::blobFromImage(make_square(image, pre_blob_image_size), blob, 1. / 255.,
                         cv::Size(INPUT_WIDTH, INPUT_HEIGHT), cv::Scalar(),
                         true, false);
  net.setInput(blob);

  /*--------------------Main Inference-----------------*/
  std::vector<cv::Mat> outputs;
  net.forward(outputs, net.getUnconnectedOutLayersNames());

  /*--------------------Post-process-------------------*/
  // max possible raw detections
  const int rows = 25200;
  // size of each raw detection
  const int dimensions = className.size() + 5;

  // Initialize vectors to hold respective outputs while unwrapping detections.
  std::vector<int> class_ids;
  std::vector<float> confidences;
  std::vector<cv::Rect> boxes;

  // start of the output data
  float *data = (float *)outputs[0].data;

  for (int i = 0; i < rows; ++i) {
    float confidence = data[4];
    // Discard bad detections and continue.
    if (confidence >= CONFIDENCE_THRESHOLD) {
      float *classes_scores = data + 5;
      cv::Mat scores(1, className.size(), CV_32FC1, classes_scores);
      // Perform minMaxLoc and acquire index of best class score.
      cv::Point class_id;
      double max_class_score;
      minMaxLoc(scores, 0, &max_class_score, 0, &class_id);
      // Continue if the class score is above the threshold.
      if (max_class_score > SCORE_THRESHOLD) {
        // Store class ID and confidence in the pre-defined respective vectors.
        confidences.push_back(confidence);

        class_ids.push_back(class_id.x);

        int left = static_cast<int>(data[0] - 0.5 * data[2]);
        int top = static_cast<int>(data[1] - 0.5 * data[3]);
        // Store good detections in the boxes vector.
        boxes.push_back(cv::Rect(left, top, static_cast<int>(data[2]), static_cast<int>(data[3])));
      }
    }

    data += dimensions;
  }

  /*--------------------Non Maximum Suppression--------------------*/
  std::vector<int> nms_result;
  cv::dnn::NMSBoxes(boxes, confidences, SCORE_THRESHOLD, NMS_THRESHOLD,
                    nms_result);

  // Resizing factor
  float resizing_factor = pre_blob_image_size / static_cast<float>(INPUT_WIDTH);
  for (int i = 0; i < nms_result.size(); i++) {
    int idx = nms_result[i];
    Detection result;
    result.class_id = class_ids[idx];
    result.confidence = confidences[idx];
    result.box.height = static_cast<int>(boxes[idx].height * resizing_factor);
    result.box.width = static_cast<int>(boxes[idx].width * resizing_factor);
    result.box.x = static_cast<int>(boxes[idx].x * resizing_factor);
    result.box.y = static_cast<int>(boxes[idx].y * resizing_factor);
    output.emplace_back(result);
  }

  // TODO: measure and store the execution time and object count of each frame.
  int inference_time = 0;
  data_record.emplace_back(inference_time);
}

/**
 * @brief This function reads the video into the RAM, placing it in a vector of frames.
 */
void loadImages(cv::VideoCapture& videoIn, std::vector<cv::Mat>& frames) {
  while (true) {
    cv::Mat frame;
    videoIn.read(frame);
    if (frame.empty()) break;
    frames.emplace_back(frame);
  }
}

int main(int argc, char **argv) {
  std::vector<std::string> class_list = load_class_list();
  std::ofstream rt_data_file("runs/data.txt",
                             std::ofstream::out | std::ofstream::trunc);
  std::cout << "This network can detect the following objects: " << std::endl;
  for (int i = 0; i < class_list.size(); ++i) {
    std::cout << class_list[i].c_str() << std::endl;
  }

  cv::VideoCapture capture("data/mario.mp4");
  if (!capture.isOpened()) {
    std::cerr << "Error opening video file\n";
    return -1;
  }

  cv::dnn::Net net;
  load_net(net, false);

  std::vector<int> data_record;

  std::vector<cv::Mat> frames;
  loadImages(capture, frames);
  size_t total_frames = frames.size();
  std::cout << "Loaded " << total_frames << " frames from the video stream.\n";

  std::vector<std::vector<int>> detection_count(total_frames);
  for (auto& d : detection_count) {
    d.resize(class_list.size(), 0);
  }

  for (size_t n = 0; n < total_frames; n++) {
    std::vector<Detection> output;
    detect(frames[n], net, output, data_record, class_list);

    // sort all detections into a histogram
    for (int i = 0; i < output.size(); ++i) {
      detection_count[n][output[i].class_id] ++;
      // For visualizing the result on a graphical system
      // cv::rectangle(frames[n], output[i].box, cv::Scalar(0, 255, 0));
    }

    // For visualizing the result on a graphical system
    /*
    cv::imshow("Detection result", frames[n]);
    cv::waitKey(1);
    */

    // TODO: count and store the numbers of other other blocks

    unsigned int time_taken_us = 0;
    std::cout << " total " << output.size() << " \tidentified objects in frame\t" << n << std::endl;
    rt_data_file << output.size() << "\tidentified objects in frame\t" << n
                 << "\tusing\t" << time_taken_us << "\tμs" << std::endl;
  }

  rt_data_file.close();
  std::cout << "Completed processing " << total_frames << " frames.\n";

  return 0;
}

File: scripts/yolo.py
import cv2
import time
import numpy as np
import sys
import os


def load_classes():
    class_list = []
    with open("network/classes_mario.txt", "r") as f:
        class_list = [cname.strip() for cname in f.readlines()]
    return class_list


def build_model(is_cuda):
    net = cv2.dnn.readNet("network/yolov5s_mario.onnx")
    if is_cuda:
        print("Attempty to use CUDA")
        net.setPreferableBackend(cv2.dnn.DNN_BACKEND_CUDA)
        net.setPreferableTarget(cv2.dnn.DNN_TARGET_CUDA_FP16)
    else:
        print("Running on CPU")
        net.setPreferableBackend(cv2.dnn.DNN_BACKEND_OPENCV)
        net.setPreferableTarget(cv2.dnn.DNN_TARGET_CPU)
    return net


colors = [(255, 255, 0), (0, 255, 0), (0, 255, 255), (255, 0, 0)]

INPUT_WIDTH = 640
INPUT_HEIGHT = 640
SCORE_THRESHOLD = 0.2
NMS_THRESHOLD = 0.4
CONFIDENCE_THRESHOLD = 0.4


def detect(image, net):
    # --------------------Pre-process--------------------
    blob = cv2.dnn.blobFromImage(
        image, 1/255.0, (INPUT_WIDTH, INPUT_HEIGHT), swapRB=True, crop=False)
    net.setInput(blob)
    # --------------------Main Inference-----------------
    preds = net.forward()
    return preds


def load_capture():
    capture = cv2.VideoCapture("data/mario.mp4")
    return capture


class_list = load_classes()


def wrap_detection(input_image, output_data):
    # --------------------Post-process-------------------
    class_ids = []
    confidences = []
    boxes = []

    rows = output_data.shape[0]

    image_width, image_height, _ = input_image.shape

    x_factor = image_width / INPUT_WIDTH
    y_factor = image_height / INPUT_HEIGHT

    for r in range(rows):
        row = output_data[r]
        confidence = row[4]
        # Discard bad detections and continue.
        if confidence >= 0.4:
            classes_scores = row[5:]
            _, _, _, max_indx = cv2.minMaxLoc(classes_scores)
            # Perform minMaxLoc and acquire index of best class score.
            class_id = max_indx[1]
            # Continue if the class score is above the threshold.
            if (classes_scores[class_id] > .25):
                # Store class ID and confidence in the pre-defined respective vectors.
                confidences.append(confidence)

                class_ids.append(class_id)

                x, y, w, h = row[0].item(), row[1].item(
                ), row[2].item(), row[3].item()
                left = int((x - 0.5 * w) * x_factor)
                top = int((y - 0.5 * h) * y_factor)
                width = int(w * x_factor)
                height = int(h * y_factor)
                # Store good detections in the boxes vector.
                box = np.array([left, top, width, height])
                boxes.append(box)

    # --------------------Non Maximum Suppression--------------------
    indexes = cv2.dnn.NMSBoxes(boxes, confidences, 0.25, 0.45)

    result_class_ids = []
    result_confidences = []
    result_boxes = []

    for i in indexes:
        result_confidences.append(confidences[i])
        result_class_ids.append(class_ids[i])
        result_boxes.append(boxes[i])

    return result_class_ids, result_confidences, result_boxes


def format_yolov5(frame):

    row, col, _ = frame.shape
    _max = max(col, row)
    result = np.zeros((_max, _max, 3), np.uint8)
    result[0:row, 0:col] = frame
    return result


# is_cuda = len(sys.argv) > 1 and sys.argv[1] == "cuda"

net = build_model(False)
capture = load_capture()
total_frames = int(capture.get(cv2.CAP_PROP_FRAME_COUNT))

start = time.time_ns()
current_frame = 0

# store data
frame_time_hist = np.zeros(shape=(total_frames, 1))
objects_detected = np.zeros(shape=(total_frames, 1))

while True:

    _, frame = capture.read()
    if frame is None:
        print("End of stream")
        break

    inputImage = format_yolov5(frame)
    outs = detect(inputImage, net)

    class_ids, confidences, boxes = wrap_detection(inputImage, outs[0])

    current_frame += 1

    print("total " + str(len(boxes)) +
          " identified objects in frame " + str(current_frame))

    # TODO: measure and store the execution time and object count of each frame.
    frame_time = 0
    frame_time_hist[current_frame - 1] = frame_time
    objects_detected[current_frame - 1] = [len(boxes)]

    # Uncomment to visualize the result on a graphical system

    # for (classid, confidence, box) in zip(class_ids, confidences, boxes):
    #     color = colors[int(classid) % len(colors)]
    #     cv2.rectangle(frame, box, color, 2)
    #     cv2.rectangle(frame, (box[0], box[1] - 20),
    #                   (box[0] + box[2], box[1]), color, -1)
    #     cv2.putText(frame, class_list[classid], (box[0],
    #                 box[1] - 10), cv2.FONT_HERSHEY_SIMPLEX, .5, (0, 0, 0))

    # cv2.imshow("output", frame)

    # if cv2.waitKey(1) > -1:
    #     print("finished by user")
    #     break

os.makedirs("runs/", exist_ok=True)
np.save('runs/frame_time.npy', frame_time_hist)
np.save('runs/objects_count.npy', objects_detected)
print("Completed processing, total frames: " + str(total_frames))
